### PR TITLE
[SPARK-36076][SQL][3.1] ArrayIndexOutOfBounds in Cast string to times…

### PR DESCRIPTION
**Pull request by dgd-contributor**
…tamp
### What changes were proposed in this pull request?
Casting string to timestamp might throw ArrayIndexOutOfBounds in certain cases
This error only occur in branch 3.0, 3.1 and previous, it's not present on 3.2 or master
Code to reproduce:
```
val df = Seq(":8:434421+ 98:38").toDF("c0")
val df2 = df.withColumn("c1", col("c0").cast(DataTypes.TimestampType))
df2.show()
```

Error:
```
java.lang.ArrayIndexOutOfBoundsException: 9
  at org.apache.spark.sql.catalyst.util.DateTimeUtils$.stringToTimestamp(DateTimeUtils.scala:328)
  at org.apache.spark.sql.catalyst.expressions.CastBase.$anonfun$castToTimestamp$2(Cast.scala:455)
  at org.apache.spark.sql.catalyst.expressions.CastBase.buildCast(Cast.scala:295)
  at org.apache.spark.sql.catalyst.expressions.CastBase.$anonfun$castToTimestamp$1(Cast.scala:451)
  at org.apache.spark.sql.catalyst.expressions.CastBase.nullSafeEval(Cast.scala:840)
  at org.apache.spark.sql.catalyst.expressions.UnaryExpression.eval(Expression.scala:476)
```

### Why are the changes needed?
Cast String to timestamp shouldn't throw error, it should return Null instead.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add test in DateTimeUtilsSuite

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@dgd-contributor can you create a PR against the master branch?

**Comment by dgd-contributor** [comment]
> @dgd-contributor can you create a PR against the master branch?
this PR is only for branch 3.0 and 3.1. On master this issue not occur since it already wrapped in a try catch

**Comment by HyukjinKwon** [comment]
@dgd-contributor can you identify the Pr that fixed this issue?

**Comment by dgd-contributor** [comment]
> @dgd-contributor can you identify the Pr that fixed this issue?

it's pull request #32936 for issue #SPARK-35720, is split func stringToTimeStamp to parseTimestampString and wrapped it in a try catch

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -237,114 +237,114 @@ object DateTimeUtils {
...
-      val parsedValue = b - '0'.toByte
-      if (parsedValue < 0 || parsedValue > 9) {
-        if (j == 0 && b == 'T') {
-          justTime = true
-          i += 3
-        } else if (i < 2) {
-          if (b == '-') {
-            if (i == 0 && j != 4) {
-              // year should have exact four digits
+    try {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:240]: Why should we make the whole block with try-catch? can we only do that for problematic block only?

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:240]: As in master branch, I see that the whole block is wrapped in, and I think we should make the whole block with try catch for better understandable and avoid corruption

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:240]: no the original function is not in try-catch. The specific function is factored out and wrapped with try-catch, isn't it?

**Review by srowen — COMMENTED** [review]
Agree, let's emulate the change in master as closely as makes sense when just 'backporting' the fix part

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 1 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:240]: yes, the factored function is wrapped with try-catch, and the factored function included the whole block and try catch is wrapped whole function

**Comment by dgd-contributor** [comment]
> Agree, let's emulate the change in master as closely as makes sense when just 'backporting' the fix part

Yes, I've emulate the fix in master, or should I put the whole code block into a function as parseTimestampString as the master does?

**Comment by srowen** [comment]
I see, it looks like both of the new functions in master are completely within try-catch and return None in case of error. If that's accurate, I think this is OK.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -237,114 +237,114 @@ object DateTimeUtils {
    *  - Region-based zone IDs in the form `area/city`, such as `Europe/Paris`
    */
   def stringToTimestamp(s: UTF8String, timeZoneId: ZoneId): Option[Long] = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:239]: Can we make the change minimised? you can do, for example, 
```suggestion
  def stringToTimestamp(s: UTF8String, timeZoneId: ZoneId): Option[Long] = try {
```

Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: ok done!

**Comment by srowen** [comment]
Jenkins test this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 3]
```diff
@@ -344,29 +344,28 @@ object DateTimeUtils {
...
     }
+    val localDateTime = LocalDateTime.of(localDate, localTime)
+    val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+    val instant = Instant.from(zonedDateTime)
+    Some(instantToMicros(instant))
+  } catch {
+    case NonFatal(_) => None
   }
 
+
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:368]: can you remove this?

**Comment by HyukjinKwon** [comment]
cc @gengliangwang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 3 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Thanks, done!

**Review by srowen — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by gengliangwang** [comment]
Merging to branch-3.1 / branch-3.0

**Comment by gengliangwang** [comment]
@dgd-contributor can you open a PR to cherry-pick to branch-3.0?

**Pull request closed without merging**